### DISCUSSION TITLE
Modbus ASCII

### DISCUSSION THREAD

**Discussion by HollyGates** [eModbus/eModbus#145] [Q&A] [answered]
Hi there, I'm looking to move to ESP32 for some of our products using an LPC micro. These products query relatively large numbers of modbus servers (up to 56) by TTL level full duplex serial (i.e. right from the UART), in combination with some MUX chips which direct the serial stream to the correct station. So I think I would need to modify the library code to set the mux lines according to modbus id before sending a query out by serial. But a bigger question is: the servers talk modbus ASCII, not RTU. Is there an easy way to configure the library to use modbus ASCII for the serial inferface?

This device would receive modbus TCP queries via a W5500 chip, and simultaneously do polling of serial connected modbus servers on a rolling basis. So its sort of like a bridge in that is acts as a TCP server, but serial client. That is what attracted me to the idea of using this library, rather than combining a TCP server library with a serial client library.

On LPC chip, all the modbus parsing and construction code was written in house, but switching to ESP32 obviously the idea of using a library is appealing. 

Thanks for any thoughts. Apologies if this is too simple or has been addressed elsewhere. I have spent some time looking at the source and examples and looked through the Discussion, but have not been able to get a sense for these questions.

**Comment by Miq1** [comment, marked as answer]
Hi, and welcome!

I never encountered a Modbus ASCII application before, so I had no urge yet to implement it as a third protocol. RTU and TCP are sharing the same inner message structure that is different from their ASCII sibling. Therefore the necessary changes would have to go very deep into the library classes, even the basic ``ModbusMessage`` class would be affected.

I am afraid that will keep me away from explicitly implementing the ASCII Modbus protocol - it would come close to a fundamental redesign, with less users requesting it.

But there may be a workaround that could be worth considering. As all Modbus ASCII messages are to start with a colon, this would be recognized in RTU as a message for server ID 58. If we had an option to treat this server ID differently, for instance dropping the CRC and adding CRLF instead, we could have a dedicated callback to deliver these messages unchecked, so that an application could construct or decode the transmitted messages at will. This way the timeout and buffer handling etc. would remain the same, but no more error checking of message contents would be possible. Of course a RTU server with ID 58 would not be possible any more at the same time...

**Reply by HollyGates** [reply]
Ok, thanks for the quick reply. It sounds like maybe it is not a great fit, so maybe I should port our internal code or else look at another library for the serial modbus side of the application.

**Reply by Miq1** [reply]
Of course, whatever suits you best. Have a successful hunt! 😁

**Reply by HollyGates** [reply]
Actually I'm coming back around to this library. I've tried a few others out there and they just don't have the flexibility to do what I need in this application. For beginners none of the other ones I looked at allow you to read or do something different for different server IDs on the modbus TCP side. The way this library is set up is also more easily fit into the way I want to have blocks of memory set up as modbus registers accessible from both the serial and the TCP side.

I'm interested in exploring your idea of doing something special with RTU server 58, if you are game!

First challenge would be getting the RTU client to send the right kind of query, which the ASCII serial server could interpret as a valid modbus ascii packet. This seems possible, as long as the library can be made to send something with an arbitrary function code and data to ID 58, as you suggested. 

On the response side, it would indeed seem like it was coming from ID 58 since the response also starts with a colon. So if the library could be made to pass that whole response on to the callback function (without failing CRC) that seems like it could work. 

Attached is an example spreadsheet with a query and response using FC03 to ASCII station #1 reading two registers back. 

I think me writing a special callback function to translate the reply message is much less work (for me) and allows me to piggyback on the excellent framework you guys have developed with the library.

One other note: I think I will need to use the blocking syncRequest calls on the serial side. The reason is that I have to switch some MUX lines in hardware before sending a query so it will go to the correct unit, then I have to wait for the response to come back before moving on to the next one (changing MUX lines).

[ASCII_RTU_EXAMPLE.xlsx](https://github.com/eModbus/eModbus/files/7945545/ASCII_RTU_EXAMPLE.xlsx)

**Comment by Miq1** [comment]
Thinking again over it,  I found another idea.
If we will restrict the communication to standard Modbus ASCII - that is, without misusing it to send plain text strings - I could have a split in the RTU send and receive functions only, leaving everything else as is.
Then one would have to use a ``useModbusASCII()`` call for the ModbusServerRTU and ModbusClientRTU instances to have the protocol switched. (And a corresponding ``useModbusRTU() `` just in case and for symmetry) 
On reception of an ASCII message it would be converted into the binary form, on sending a binary would be converted into ASCII again.
There might be additional error codes for ASCII frame errors etc., but the majority of the code can remain as is.
How do you rate that?

**Reply by HollyGates** [reply]
That sounds great! A translator function is a good approach since it is pretty easy to convert ASCII to RTU and back, then all the core library stuff can stay the way it is.

I am sorry I am not at a high enough level to contribute to the library in a significant way and try to make structural changes myself. I have a lot of experience with firmware in C, but almost none with C++. Only experience with object oriented was Java in college years ago, so I am still getting used to the way C++ code looks and works. Plus I'm not really a great programmer anyway, I am stronger in hardware design. 

But I am happy to beta test things, give feedback, and contribute example code if it is useful. I appreciate all your work to create and support this library.

**Reply by Miq1** [reply]
No worries, I am in charge anyway 🤣 

I cannot give you a time frame currently, but I promise to work on it in the next few days. I would appreciate if you could do some testing once I have a working code. I probably will open another branch for it - let us say: ModbusASCII? - that you will be required to download in case.

There should be no difference at all on the application side, except the need to call ``useModbusASCII()`` once.

**Reply by HollyGates** [reply]
Sure, I'm happy to do testing. Next week I will set up some hardware on my desk to attach a single serial server at ID 16 to an ESP32 board, with logic analyzer on TX/RX so I can see what is going between the two. Then when you have a chance to work on that code I can be in a position to test on real hardware.

In my normal application I continuously read 10 holding registers starting at address 0 using FC03. So maybe we should work on a modified version of RTU04example?

The first thing I'm working on just reads analog values and is a TCP server that gets read by the PLC, so the ascii serial client thing is not as urgent. But that is the second application I want to convert to ESP32 so I'm definitely interested in it and happy to work with you to develop the ASCII capability.

**Reply by Miq1** [reply]
👍🏻 
I will drop you a note here as soon as you may start off.

**Reply by HollyGates** [reply]
Awesome, thanks very much.

**Comment by Miq1** [comment]
I committed a first attempt on ModbusASCII in the said branch. It compiles, but is completely untested yet. I left the timing functionalities of RTU still active, I might have to remove those for ASCII mode as well.

The only thing you need to do after defining a ``ModbusServerRTU`` or ``ModbusClientRTU`` to switch it is to call its ``void useModbusASCII()`` method. You may go back by ``void useModbusRTU()`` and check the current state with ``bool isModbusASCII()``.

Please note that already enqueued requests will definitely result in errors after the switch, since the mode is held with the server/client, not with the respective message.

Next I will extend the test sketch to do some repeatable test cases.

**Reply by HollyGates** [reply]
Excellent. I am planning on getting some hardware set up next week to test this out. I also need to spend some time getting familiar with how the RTU client setup works; so far I've just been working with TCP server. An example would definitely help me get there faster.

**Comment by Miq1** [comment]
Uh,  wait a bit before using the ASCII code. I noticed the timeout handling in the RTU receive loop in fact is not working well with ASCII messages. I need to implement a second state machine here. PLS SBY!

**Comment by Miq1** [comment]
New version up. Again untested here, but compiling okay.

Changes:
- dropped previous attempt to handle ASCII in the RTU ``receive()``, but built a separate state machine for ASCII. Now the 1s-Timeout (or whatever you will set it to **after** calling ``useModbusASCII()``  for the client/server) is handled independently from RTU timings
- Made the lead-in and lead-out sequences less flexible, now using fixed ``:<message><crc>\r\n`` form. The previous wa overkill...
- Do CRC calculation while receiving, no separate call any more.

Give it a try if you find the time.

**Reply by HollyGates** [reply]
Realized I was missing a chip for building my test setup. Ordered from Digikey but I won't be able to build it for a few more days. Any chance of you putting together a skeleton example main.cpp of how to use your new ascii features?

**Reply by Miq1** [reply]
Take your time. I will first add some ASCII mode test cases to the test sketch in the next days to see if it is working as I intended. May be I will do an example later on as well, but that will be not much different from any RTU example, just having the one additional call to ``useModbusASCII()`` in ``setup()``.

**Comment by Miq1** [comment]
The CRC checks were done doubly (ASCII **and** RTU CRCs...), so I had to reconstruct the mechanism a bit. New commit will be up in a minute.

My test sketch is funnily senseless by the way, as the changes only do affect the lower protocol levels that are not seen at all on application level... 😜

**Reply by HollyGates** [reply]
I had a chance to try this out today. First test is close but needs a small tweak. 

First I recorded with analyzer the normal ASCII interchange between client and server with the logic analyzer. In this case the client is a Lantronix XPORT ethernet bridge module on a base PCB I have which interfaces electrically and mechancially with the modbus server device. The physical layer for that serial interface is NOT RS485, but just regular 3.3V async serial straight out of UARTs. Here is what I saw with an FC03 read of station 16 (what this device is set to), 2 registers starting at 0.

![mbASCII_ltrx1](https://user-images.githubusercontent.com/4689112/152426754-7cc0f22f-2d51-4094-aeb6-9f92991576a2.PNG)

The client sends :100300000002EB (plus \r\n)
The server responds with :10030400040000E5 (plus double \r\n, which is weird, I'll have to look at that at some point)

This device is a two color pyrometer which optically measures temps of graphite objects inside our wafer making furnace. There are many of these on one furnace.

Next I wired up an ESP32 HUZZAH board to a base PCB I had modified by removing the Lantronix bridge module, connecting to the RX/TX pins on the HUZZAH, connected to Serial2 in code. 

![image](https://user-images.githubusercontent.com/4689112/152427274-9b33dab2-9eb1-4d86-b91f-dfc0700ebf73.png)

I made a program to just Serial2.println the query listed above and looked with logic analyzer. After getting a few things straightened out, this worked and gave the same logic analyzer trace included above.

Next, I did the RTU04exampe from the eModbus master branch project. I adjusted a few things but pretty much like the example in the repository. This gave the me logic trace:

![image](https://user-images.githubusercontent.com/4689112/152427804-7a24f5dd-0ea1-4c42-9e36-17268150b217.png)

This looks like what I would expect for RTU. Of course there is no answer from the device since it only talks ASCII!

Next I changed the platformio.ini to pull the ASCII branch

![image](https://user-images.githubusercontent.com/4689112/152427977-68bfe655-391a-42bb-af1a-935c8c85c3d0.png)

And added a line to switch the modbus flavor to ascii. Here is what I get on the logic analyzer:

![image](https://user-images.githubusercontent.com/4689112/152428147-06deafab-b993-4d2e-9e34-52f75adeb5dc.png)

So the LRC is EA when it should be EB. I changed some options to match an example case from here:

![image](https://user-images.githubusercontent.com/4689112/152428304-381d9eb6-3708-4a1c-940c-f19d1dbe2b40.png)

But I got this:

![image](https://user-images.githubusercontent.com/4689112/152428371-ad78545e-6350-4198-ae64-25420e390914.png)

So it appears the LRC is off by one. I looked at the code in RTUutils.cpp. The LRC is supposed to be 2's complement of the rolling 8 bit sum of the message characters. Here is the relevant code:
![image](https://user-images.githubusercontent.com/4689112/152438902-b9479cb9-a1f5-496c-87d9-bd113a8f77f7.png)

The inversion is done for twos complement, but not the plus one. So I added one line:
![image](https://user-images.githubusercontent.com/4689112/152439132-015bf741-0b77-49aa-8d16-76852c80770f.png)

Now it is sending the right thing, and I see a response from the device:

![image](https://user-images.githubusercontent.com/4689112/152439875-e36541d4-c8d6-4c5f-931d-bf5ed59fedd3.png)

But on the terminal I see that ModbusClientRTU doesn't like the CRC (i.e. LRC):

```
[D] 6088| ModbusClientRTU.cpp  [ 198] handleConnection: Request sent.
[D] Raw buffer received: @3FFB2068/8:
  | 0000: 10 03 04 00 04 00 00 E5                           |........        |
[D] Received packet: @3FFBDB4C/1:
  | 0000: EE                                                |.               |
[D] 6114| ModbusClientRTU.cpp  [ 204] handleConnection: Error response (1 bytes) received.
[D] 6122| ModbusClientRTU.cpp  [ 225] handleConnection: Response generated.
␛[1;33m[E] 6128| main.cpp             [  58] handleError: Error response: EE - Invalid ASCII CRC
```

I thought probably the same bug for LRC check on reply packet. Also in RTUutils.cpp:

![image](https://user-images.githubusercontent.com/4689112/152440707-f65ec092-ee60-4bf7-b5c1-f2d3aa084ffc.png)

I changed to:

![image](https://user-images.githubusercontent.com/4689112/152440750-6974a3e2-e223-4c12-961d-bff0426d56e9.png)

Now I get a receive in terminal!

```
[D] 715293| ModbusClientRTU.cpp  [ 126] addRequestM: request for 10/03
[D] 715293| ModbusClientRTU.cpp  [ 175] addToQueue: RC=01
[D] 715293| ModbusClientRTU.cpp  [ 137] addRequestM: RC=00
[D] 715293| ModbusClientRTU.cpp  [ 193] handleConnection: Pulled request from queue
[D] Sent packet: @3FFB215C/6:
  | 0000: 10 03 00 00 00 02                                 |......          |
[D] 715315| ModbusClientRTU.cpp  [ 198] handleConnection: Request sent.
[D] Raw buffer received: @3FFB21BC/8:
  | 0000: 10 03 04 00 04 00 00 E5                           |........        |
[D] Received packet: @3FFB215C/7:
  | 0000: 10 03 04 00 04 00 00                              |.......         |
[D] 715341| ModbusClientRTU.cpp  [ 204] handleConnection: Data response (7 bytes) received.
[D] 715349| ModbusClientRTU.cpp  [ 225] handleConnection: Response generated.
Requested at  715.293s:
   0000:    0.000
   0002:    0.000
----------
```

Nice! However, data in register 0 should be 4... have to look into that tomorrow.

Thanks for adding this feature! I'm excited to put the library into use in our solar wafer making machines.






```

**Reply by HollyGates** [reply]
Nvm, the data in handler was interpreting as floats so I changed it to int16_t and now it reads the right thing for my device.

```
[D] 3053| ModbusClientRTU.cpp  [ 126] addRequestM: request for 10/03
[D] 3053| ModbusClientRTU.cpp  [ 175] addToQueue: RC=01
[D] 3053| ModbusClientRTU.cpp  [ 137] addRequestM: RC=00
[D] 3053| ModbusClientRTU.cpp  [ 193] handleConnection: Pulled request from queue
[D] Sent packet: @3FFB99BC/6:
  | 0000: 10 03 00 00 00 02                                 |......          |
[D] 3074| ModbusClientRTU.cpp  [ 198] handleConnection: Request sent.
[D] Raw buffer received: @3FFB2068/8:
  | 0000: 10 03 04 00 04 00 00 E5                           |........        |
[D] Received packet: @3FFB99BC/7:
  | 0000: 10 03 04 00 04 00 00                              |.......         |
[D] 3100| ModbusClientRTU.cpp  [ 204] handleConnection: Data response (7 bytes) received.
[D] 3108| ModbusClientRTU.cpp  [ 225] handleConnection: Response generated.
Requested at    3.053s:
   0000: 4
   0001: 0
----------
```

One problem is that if the device times out, the watchdog resets the ESP32:

```
E (12239) task_wdt: Task watchdog got triggered. The following tasks did not reset the watchdog in time:
E (12239) task_wdt:  - IDLE0 (CPU 0)
E (12239) task_wdt: Tasks currently running:
E (12239) task_wdt: CPU 0: Modbus01RTU
E (12239) task_wdt: CPU 1: loopTask
E (12239) task_wdt: Aborting.
abort() was called at PC 0x400e42bf on core 0

ELF file SHA256: 0000000000000000

Backtrace: 0x40084f3c:0x3ffbe500 0x400851b1:0x3ffbe520 0x400e42bf:0x3ffbe540 0x40083971:0x3ffbe560 0x4000bfed:0x3ffb9450 0x40087025:0x3ffb9460 0x40086147:0x3ffb9480 0x400d41e9:0x3ffb94a0 0x400d35ed:0x3ffb94c0 0x400d32ba:0x3ffb94e0 0x400d240d:0x3ffb9530 0x400861c1:0x3ffb95e0

Rebooting...
ets Jul 29 2019 12:21:46
```

**Comment by Miq1** [comment]
Great you got it basically working.
I coded the LRC calculation as I thought could be right, based on some description I found on the net. There it said when checking messages the addition including the LRC should sum up to 0 - but I got 0xFF all the time, so I made the check against that. The +1 bit was never mentioned.
Anyway, I will add the lines as you found were working okay.
The WDT reset almost was to be expected,  since there is no ``delay()`` in the state machine yet. So it will run a second in a closed loop waiting for data without yielding to the task manager. I will fix that as well - today, most probably,

**Comment by Miq1** [comment]
I fixed the issues you mentioned - another try is appreciated 😉

**Comment by hgatesCPV** [comment]
Great, LRC thing is good now. Also no rebooting with device disconnected.

Some more questions for you. In this particular application I want to scan through a bunch of serial devices as quickly as possible using Client, storing the results in a register bank. That block of data is then pulled up to the PLC over ethernet by TCP server. In one board, I have 56 devices and we would like to poll them all in 500 or 1000ms. The basic query and response takes about 4ms. 

In the demo code, I turned down the READ_INTERVAL lower and lower until I started getting problems. As it got down around 10ms, it would work for about one second but then start printing errors that the new request could not be added because the queue was full. That makes sense to me; if I am adding requests faster than responses are coming back eventually the queue will get full.

I modified the code to wait until data is available before issuing another request, so that only a single one is active at a time, then measured what frequency the requests were coming out at. See logic analyzer trace below. There is about 8ms between the end of one response and the start of the new query. Do you have any ideas about why this time is so long, or what could be done to speed it up? I could probably live with it, but faster would be better. This is with all serial terminal printing disabled, so I don't think that is slowing anything down.

![image](https://user-images.githubusercontent.com/99047549/152597588-c80c3961-f09a-4663-9086-af2bb2d1dc51.png)

Another question is on timeouts. I tried setting the timeout to 100ms, and on startup it prints that timeout is set to 100. However, the queries only come out at 1s intervals. This is with the code that makes a new request as soon as the last one either gets data or an error.

![image](https://user-images.githubusercontent.com/99047549/152597823-8f45bed9-a875-4a36-9de0-66d237280e71.png)

I looked around and found this:
![image](https://user-images.githubusercontent.com/99047549/152603076-c0f38eaf-ac96-4357-8d4b-f556107a46e7.png)

I would maybe not set to 1s for ASCII - the Lantronix bridges are highly configurable this way and we usually set them to 200ms timeout. We want a low timeout because if you have for instance one of those 56 devices removed for cleaning (this happens), you don't want to mess up scanning the other 55 of them while the one is removed.

But anyway, if I just set the timeout AFTER setting ascii mode and that works fine

```C++
// Set up Serial2 connected to Modbus RTU
  Serial2.begin(BAUDRATE, SERIAL_8N1, RXPIN, TXPIN);

  // Set up ModbusRTU client.
  // - provide onData handler function
  MB.onDataHandler(&handleData);
  // - provide onError handler function
  MB.onErrorHandler(&handleError);
  MB.useModbusASCII();
  MB.setTimeout(100);
  // Start ModbusRTU background task
  MB.begin();
```

More testing later...

**Reply by HollyGates** [reply]
Oops, used my other account to post the above by accident.

**Comment by Miq1** [comment]
The 1s timeout is defined by ModbusASCII - if there is no character received within that time the message shall be considered faulty. Of course you may override that exactly the way you chose. I am wondering why the 1s should apply if the message has been received completely including the trailing CRLF. I thought I had coded it to proceed right away. I will have to look at it again tomorrow.
The 8ms between requests may be the fastest the code is able to achieve ; there is quite some stuff running in between to clean up the queues etc., especially when you are using the synchronous calls.

**Reply by Miq1** [reply]
Oops. Had a quick look at the code and found the timeout handling is missing a relevant point : when a byte is read successfully, the ``TimeOut`` variable needs to be set to ``millis() `` again.
Will fix this tomorrow.

**Comment by HollyGates** [comment]
The 1s doesn't apply if I set timeout after setting ascii mode. It is about 110ms, which is close enough to the 100ms I asked for.

I am using the non-blocking ones for now, just keeping an eye on when the request gets answered, then making another request immediately. 

```C++

uint8_t open_requests = 0;

void handleError(Error error, uint32_t token)
{
  // ModbusError wraps the error code and provides a readable error message for it
  ModbusError me(error);
  LOG_E("Error response: %02X - %s : open: %u\r\n", (int)me, (const char *)me, open_requests);
  if (open_requests > 0)
    open_requests--;
}

void loop()
{
  uint32_t token;
  // do another request?
  if (open_requests == 0)
  {
    token = (uint32_t)millis();
    Error err = MB.addRequest(token, 16, READ_HOLD_REGISTER, FIRST_REGISTER, NUM_VALUES);
    if (err != SUCCESS)
    {
      ModbusError e(err);
      LOG_E("Error creating request: %02X - %s : open: %u\r\n", (int)e, (const char *)e, open_requests);
    }
    else
      open_requests++;
  }
  // No, but we may have another response
  if (data_ready)
  {
    // We do. Print out the data
    // Serial.printf("got response to request at %8.3fs : open:%u\r\n", request_time / 1000.0, open_requests);
    // for (uint8_t i = 0; i < NUM_VALUES; ++i)
    //{
    //   Serial.printf("   %04X: %i\n", i + FIRST_REGISTER, values[i]);
    //}
    // Serial.printf("----------\n\n");
    data_ready = false;

    if (open_requests > 0)
      open_requests--;
  }
}
```

**Reply by Miq1** [reply]
Alternatively you could use the queue size parameter to increase the client queue to the number of requests + x you are expecting to send in one burst. The client code then will do exactly that: send another request as soon as the previous response has been received.
The handleData callback must store away the received data as quickly as possible to speed up the cycle.
But keep in mind that the handleData and handleError callbacks are run on the client core, so to further process the data in loop you will need to protect it by mutex.

**Reply by Miq1** [reply]
> The 1s doesn't apply if I set timeout after setting ascii mode. It is about 110ms, which is close enough to the 100ms I asked for.

The additional 10ms will add in as soon as the previous check on timeout was just below the set value. To keep the WDT happy I am adding 10ms delay whenever no character was seen.

**Comment by Miq1** [comment]
I will make the timeout an optional parameter for ``useModbusASCII()`` that defaults to 1s. So you may alter it for your purposes without another call.

**Comment by Miq1** [comment]
Changes are up now. There is the announced optional timeout parameter in the ``useModbusASCII()`` calls, the timeout handling has been fixed and I made the character checks in ASCII mode somewhat more efficient.

**Reply by hgatesCPV** [reply]
Just tested your latest - looks good. I tried the new timeout parameter and it works as expected. No change in timing seen for max speed polling demo; still about 8ms between end of response and beginning of new query. Anyway, I think this is good for now. I need to test FC23 and some other things but I don't have hardware set up for that at the moment.

**Reply by Miq1** [reply]
I doubt we will be able to go below the 8ms without massive design changes. There are some ``delay(1)`` lines to give the task scheduler a chance to work that we cannot further reduce. I think that will be part of the time spent, another is the queue handling. I do not see an easy way to streamline that, though.

**Reply by Miq1** [reply]
You might try once more. I moved one delay in RTU maode to another spot, hit less often. I would be interested if that changes a bit from the 8ms. It is increasing the risk that the WDT will strike, though,  and will reset the ESP...

**Reply by hgatesCPV** [reply]
Wow, much faster. Delay is now down to about 1.4mS so full query/response cycle time is mostly 5ms, occasionally 6ms. In a short amount of testing I am not seeing reboots, but I will let it run for longer and see. Thanks! This is awesome.

![image](https://user-images.githubusercontent.com/99047549/152860907-e77b15f5-de06-4884-ab89-ae1058583a2a.png)

**Reply by Miq1** [reply]
Thanks for testing and I am glad to hear that it seems to be working for you.
The risk of resets is directly dependent from the performance of the servers. If we happen to have a laggy one, the reset will be much more likely.
Maybe I will have to make that a configuration option.
